contracts/lens/Quoter.sol

```diff
@@ -0,0 +1,320 @@
+        private
+        pure
+        returns (
+            int128[] memory deltaAmounts,
+            uint160[] memory sqrtPriceX96AfterList,
+            uint32[] memory initializedTicksLoadedList
+        )
+    {
+        reason = validateRevertReason(reason);
+        (deltaAmounts, sqrtPriceX96AfterList, initializedTicksLoadedList) =
+            abi.decode(reason, (int128[], uint160[], uint32[]));
+    }
+
+    /// @dev quote an ExactInput swap along a path of tokens, then revert with the result
+    function _quoteExactInput(QuoteExactInputParams memory params) public selfOnly returns (bytes memory) {
+        uint256 pathLength = params.path.length;
+
+        int128[] memory deltaAmounts = new int128[](pathLength + 1);
+        uint160[] memory sqrtPriceX96AfterList = new uint160[](pathLength);
+        uint32[] memory initializedTicksLoadedList = new uint32[](pathLength);
+        Currency prevCurrencyOut;
+        uint128 prevAmountOut;
+
+        for (uint256 i = 0; i < pathLength; i++) {
+            (PoolKey memory poolKey, bool zeroForOne) =
+                PathKeyLib.getPoolAndSwapDirection(params.path[i], i == 0 ? params.currencyIn : prevCurrencyOut);
+            (, int24 tickBefore,) = manager.getSlot0(poolKey.toId());
+
+            (BalanceDelta curDeltas, uint160 sqrtPriceX96After, int24 tickAfter) = _swap(
+                poolKey,
+                zeroForOne,
+                int256(int128(i == 0 ? params.amountIn : prevAmountOut)),
+                0,
+                params.path[i].hookData
+            );
+
+            (int128 deltaIn, int128 deltaOut) =
+                zeroForOne ? (curDeltas.amount0(), curDeltas.amount1()) : (curDeltas.amount1(), curDeltas.amount0());
+            deltaAmounts[i] += deltaIn;
+            deltaAmounts[i + 1] += deltaOut;
+
+            prevAmountOut = zeroForOne ? uint128(-curDeltas.amount1()) : uint128(-curDeltas.amount0());
+            prevCurrencyOut = params.path[i].intermediateCurrency;
+            sqrtPriceX96AfterList[i] = sqrtPriceX96After;
+            initializedTicksLoadedList[i] =
+                PoolTicksCounter.countInitializedTicksLoaded(manager, poolKey, tickBefore, tickAfter);
+        }
+        bytes memory result = abi.encode(deltaAmounts, sqrtPriceX96AfterList, initializedTicksLoadedList);
+        assembly {
+            revert(add(0x20, result), mload(result))
+        }
+    }
+
+    /// @dev quote an ExactInput swap on a pool, then revert with the result
+    function _quoteExactInputSingle(QuoteExactInputSingleParams memory params) public selfOnly returns (bytes memory) {
+        (BalanceDelta deltas, uint160 sqrtPriceX96After, int24 tickAfter) = _swap(
+            params.poolKey,
+            params.zeroForOne,
+            int256(int128(params.amountIn)),
+            params.sqrtPriceLimitX96,
+            params.hookData
+        );
+        bytes memory result = abi.encode(deltas, sqrtPriceX96After, tickAfter);
```

Comment:
in the _quoteExactInput logic we directly encode the values we want in _handleRevert. Why don't we do the same for _quoteExactInputSingle / _handleRevertSingle? ie push the initializedTicksLoaded call and encoding the deltaAmounts as an array into _quoteExactInputSingle ?